### PR TITLE
VPN Control Buttons in Systray

### LINKED ISSUE #860: Improve Systray App
**Feature description**

The systray application currently does not support:

- stopping/starting the VPN 
- show VPN status (off/on/dialing)
- quick selection of VPN servers

It also does grey out the `uninstall` option if the Systray app is run manually (not package based). 

These features would be valuable for users. 

Concrete wireframes for these features are yet to be done.

### PR DESCRIPTION
**Pull request by mrpalide**
Did you run `make format && make check`? Yes

Fixes #860 

 Changes:	
- Add VPN control button on systray:
  - VPN connection status
  - Select server from list
  - Connect/Disconnect button
- Remove transport cache and get it before create route

How to test this PR:
- Needs to install systray dependencies, follow instruction on [the docs](https://github.com/mrpalide/skywire/blob/feature/improve-systray/docs/systray-builds.md)
- `make build-systray` [or `make build-systray-windows` for Windows]
- Run visor
  - **note:** In Windows machine visor should run as administrator to make routes for VPN client. But in *nix machine, please set `cap_net_admin` by `sudo setcap 'cap_net_admin+p' ./apps/vpn-client` command and run visor normal (without root permission).
- Run VPN server on test env, and choose it from list and use/test other buttons

### CODE REVIEW

**Comment by ersonp** [comment]
Getting this on windows, yet to test on linux
```
2022/03/08 18:30:24 [DESKTOP-J9DSNEK/iZzZ4H97Dz-000004] "GET http://localhost:8000/api/health HTTP/1.1" from [::1]:50461 - 404 19B in 352.3µs
```

File: internal/gui/gui.go [hunk 1]
```diff
@@ -168,13 +190,102 @@ func initOpenVPNLinkBtn(vc *visorconfig.V1) {
...
+	go serversBtn(conf, mVPNServers, rpcClient)
+}
+
+func vpnStatusBtn(conf *visorconfig.V1, rpcClient visor.API) {
+	for {
+		stats, _ := rpcClient.GetAppConnectionsSummary(skyenv.VPNClientName)
+		if len(stats) == 1 {
+			if stats[0].IsAlive {
+				mVPNStatus.SetTitle("Status: Connected")
+				mVPNButton.SetTitle("Off")
```
Comment (ersonp) [inline internal/gui/gui.go]: This should be `Dissconnect`

File: internal/gui/gui.go [hunk 2]
```diff
@@ -168,13 +190,102 @@ func initOpenVPNLinkBtn(vc *visorconfig.V1) {
...
+				mVPNButton.Enable()
+			} else {
+				mVPNStatus.SetTitle("Status: Connecting...")
+				mVPNButton.SetTitle("Off")
```
Comment (ersonp) [inline internal/gui/gui.go]: And this `Connect`

**Comment by ersonp** [comment]
```
 panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                           
 [signal 0xc0000005 code=0x0 addr=0x0 pc=0xf1eda4]                                                                                                                                                                                                                                                                                                                                                                                                                   
 goroutine 203 [running]:                                                                                                                                                                                                          
 github.com/skycoin/skywire/internal/gui.stopVisor()                                                                                                                                                                                       
 D:/Code/go/skywire/internal/gui/gui.go:448 +0x44                                                                                                                                                                          
 github.com/skycoin/skywire/internal/gui.Stop()                                                                                                                                                                                            
 D:/Code/go/skywire/internal/gui/gui.go:122 +0x32                                                                                                                                                                          
 github.com/skycoin/skywire/internal/gui.handleUserInteraction(0xdd1ee6, 0xc000289b00)                                                                                                                                                     
 D:/Code/go/skywire/internal/gui/gui.go:413 +0x13b                                                                                                                                                                         
 created by github.com/skycoin/skywire/internal/gui.GetOnGUIReady.func1                                                                                                                                                                    
 D:/Code/go/skywire/internal/gui/gui.go:91 +0x125    
```
On using `Quit` from systray

**Comment by ersonp** [comment]
`make build-systray-windows` builds apps without `.exe` extensions so it gives `file does not exist` error. Apps build from `make build-windows` work.

```
[2022-03-08T19:42:03+05:30] INFO [visor]: Changing vpn-client PK to "03af84a974e3f89a50c0f7f3e13507bc814127e37d3736cebaea59fa003a726dd5"
[2022-03-08T19:42:03+05:30] INFO [visor:config]: Flushing config to file. config_version="v1.1.1" filepath=".\skywire-config.json"
[2022-03-08T19:42:03+05:30] INFO [visor]: Updated vpn-client PK
[2022-03-08T19:42:03+05:30] INFO [visor_rpc:CLI]: Request processed. _elapsed="509.5µs" _method="SetAppPK" _received="7:42PM" input=&{AppName:vpn-client PK:03af84a974e3f89a50c0f7f3e13507bc814127e37d3736cebaea59fa003a726dd5}
[2022-03-08T19:42:04+05:30] INFO [visor_rpc:CLI]: Request processed. _elapsed="606.4224ms" _method="StartApp" _received="7:42PM" error="exec: "D:\\Code\\go\\skywire\\apps\\vpn-client": file does not exist" input=0xc00024bce0
[2022-03-08T19:42:05+05:30] INFO [visor_rpc:CLI]: Request processed. _elapsed="0s" _method="GetAppConnectionsSummary" _received="7:42PM" error="no such app" input=0xc0003468b0 output=&[]
```

with apps built from `make build-systray-windows`
```
[2022-03-08T19:45:25+05:30] INFO [visor]: Changing vpn-client PK to "03af84a974e3f89a50c0f7f3e13507bc814127e37d3736cebaea59fa003a726dd5"
[2022-03-08T19:45:25+05:30] INFO [visor:config]: Flushing config to file. config_version="v1.1.1" filepath=".\skywire-config.json"
[2022-03-08T19:45:25+05:30] INFO [visor]: Updated vpn-client PK
[2022-03-08T19:45:25+05:30] INFO [visor_rpc:CLI]: Request processed. _elapsed="1.026ms" _method="SetAppPK" _received="7:45PM" input=&{AppName:vpn-client PK:03af84a974e3f89a50c0f7f3e13507bc814127e37d3736cebaea59fa003a726dd5}
[2022-03-08T19:45:25+05:30] INFO [visor_rpc:CLI]: Request processed. _elapsed="387.5145ms" _method="StartApp" _received="7:45PM" input=0xc000284000
[2022-03-08T19:45:26+05:30] INFO (STDOUT) [proc:vpn-client:470c362d9ea04a8eba51fd7195423bbe]: Connecting to VPN server 03af84a974e3f89a50c0f7f3e13507bc814127e37d3736cebaea59fa003a726dd5
[2022-03-08T19:45:26+05:30] INFO [proc_manager]: Accepted proc conn. hello="{"proc_key":"470c362d9ea04a8eba51fd7195423bbe","egress_net":"tcp","egress_addr":"127.0.0.1:58648","event_subs":{"tcp_close":true,"tcp_dial":true}}" remote=127.0.0.1:58650
```

with apps built from `make build-windows`

File: internal/gui/gui.go (hunk 1 above)
Comment (mrpalide) [reply internal/gui/gui.go]: Done.

File: internal/gui/gui.go (hunk 2 above)
Comment (mrpalide) [reply internal/gui/gui.go]: Done.

**Comment by ersonp** [comment]
Running the systray on ubuntu linux has some issues
```
sudo ./skywire-visor -c skywire-config.json --systray
```
Running it with sudo doesn't give any error logs but nothing can be seen on systray. But on shutdown this warning is seen
```
[2022-03-09T14:01:48+05:30] INFO [2/12] [visor:shutdown:cli.listener]: Module stopped cleanly. elapsed=4.710203ms
[2022-03-09T14:01:48+05:30] INFO [1/12] [visor:shutdown:event_broadcaster]: Shutting down module...
[2022-03-09T14:01:48+05:30] INFO [1/12] [visor:shutdown:event_broadcaster]: Module stopped cleanly. elapsed=4.559321ms
[2022-03-09T14:01:48+05:30] INFO [visor:shutdown]: Shutdown complete. Goodbye!

(skywire-visor:7398): libayatana-appindicator-WARNING **: 14:01:48.575: Unable to send signal for NewStatus: The connection is closed
```

And when run without sudo
```
./skywire-visor -c skywire-config.json --systray
```
The this log is seen and the visor is stuck in the loop even after stopping the visor with `Ctrl+C`
```
[2022-03-09T14:05:03+05:30] INFO [1/12] [visor:shutdown:cli.listener]: Shutting down module...
2022/03/09 14:05:03 rpc.Serve: accept:accept tcp 127.0.0.1:3435: use of closed network connection
[2022-03-09T14:05:03+05:30] INFO [1/12] [visor:shutdown:cli.listener]: Module stopped cleanly. elapsed=53.279µs
[2022-03-09T14:05:03+05:30] INFO [visor:shutdown]: Shutdown complete. Goodbye!
[2022-03-09T14:05:04+05:30] INFO [systray:rpc_client]: Waiting for RPC get ready...
[2022-03-09T14:05:06+05:30] INFO [systray:rpc_client]: Waiting for RPC get ready...
[2022-03-09T14:05:08+05:30] INFO [systray:rpc_client]: Waiting for RPC get ready...
```

**Comment by 0pcom** [comment]
instead of the icon being defined in `/opt/skywire/icon.png` for linux, I think it should be set to:
`/usr/share/icons/hicolor/48x48/apps/skywire.png`

that is where it currently is in the packages.

**Comment by 0pcom** [comment]
image

The icon briefly appears in the systray and then disappears when this crashes

on archlinux / endeavourOS

File: internal/gui/gui.go [hunk 3]
```diff
@@ -168,13 +182,111 @@ func initOpenVPNLinkBtn(vc *visorconfig.V1) {
...
 
-//func initVpnClientBtn() {
-//	mVPNClient = systray.AddMenuItem("VPN", "VPN Client Connection")
-//}
+func initVpnClientBtn(conf *visorconfig.V1, httpClient *http.Client, logger *logging.MasterLogger) {
+
+	rpc_logger := logger.PackageLogger("systray:rpc_client")
+	hvAddr := getHVAddr(conf)
+	for !isHypervisorRunning(hvAddr) {
+		rpc_logger.Info("Waiting for RPC get ready...")
```
Comment (ersonp) [inline internal/gui/gui.go]: `Waiting for RPC to get ready...`

**Review by ersonp — APPROVED** [review]
Working on Linux and Windows.

**Comment by 0pcom** [comment]
Works for mac

image

**Review by 0pcom — APPROVED** [review]

**Pull request merged**